hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java

```diff
@@ -707,20 +709,37 @@ protected List<String> getInstantsToRollback(HoodieTableMetaClient metaClient, H
         }
       }).collect(Collectors.toList());
     } else if (cleaningPolicy.isLazy()) {
-      return inflightInstantsStream.filter(instant -> {
-        try {
-          return heartbeatClient.isHeartbeatExpired(instant.getTimestamp());
-        } catch (IOException io) {
-          throw new HoodieException("Failed to check heartbeat for instant " + instant, io);
-        }
-      }).map(HoodieInstant::getTimestamp).collect(Collectors.toList());
+      return getInstantsToRollbackForLazyCleanPolicy(metaClient, inflightInstantsStream);
     } else if (cleaningPolicy.isNever()) {
       return Collections.emptyList();
     } else {
       throw new IllegalArgumentException("Invalid Failed Writes Cleaning Policy " + config.getFailedWritesCleanPolicy());
     }
   }
 
+  @VisibleForTesting
+  public List<String> getInstantsToRollbackForLazyCleanPolicy(HoodieTableMetaClient metaClient,
+                                                              Stream<HoodieInstant> inflightInstantsStream) {
+    // Get expired instants, must store them into list before double-checking
+    List<String> expiredInstants = inflightInstantsStream.filter(instant -> {
+      try {
+        // An instant transformed from inflight to completed have no heartbeat file and will be detected as expired instant here
+        return heartbeatClient.isHeartbeatExpired(instant.getTimestamp());
+      } catch (IOException io) {
+        throw new HoodieException("Failed to check heartbeat for instant " + instant, io);
+      }
+    }).map(HoodieInstant::getTimestamp).collect(Collectors.toList());
+
+    if (!expiredInstants.isEmpty()) {
```

Comment:
generally reloading entire timeline is a costly operation. I am trying to see if there is any other elegant way to fix this.

Ideally the heart beat file should always be generated by the regular writer. and hence only condition when the heart beat file goes missing is, the writer completed the commit and deleted it.  and if thats a safe assumption, if we encounter heart beat file not found, we should ignore the commit of interest. 

Only catch here is. if a user switches from non lazy to lazy clean, those may not have heart beats. 

Can we introduce locking for entire getInstantsToRollback. and so that we can guarantee that if a writer is concurrently completing a commit, we can deduce that when compared to a failed commit in previous launch when writer had eager failed policy.

Reply:
or we can also try checking for existence of completed commit file in ".hoodie" after we get hold of list of commits to rollback. Those which seem to have a completed instant, we can ignore them from the list. 
let me know what do you think.

Reply:
I think checking for existence of completed commit file in ".hoodie"  is much better than reloading the whole timeline, which is very lightweight while introducing a lock for entire getInstantsToRollback may be kind of complicated.  I will push a new commit that only check the existence of commit file instead of reloading the timeline.

Reply:
@nsivabalan Hi,  all comments are addressed. Now we will check the existence of commit file instead of reloading the timeline when doing the double check